File: PerformanceTest/swimmingpool.js
/*
 * Swimming Pool Simulation (JavaScript) - WITH LOGGING
 * ====================================================
 * This script simulates a swimming pool environment using the SimLuxJS framework.
 * It models customer arrivals, swimming durations, and pool capacity constraints.
 * With configurable logging modes, it allows for testing and debugging.
 *
 * LOGGING MODES:
 * - 'console': Real-time output to terminal
 * - 'file': Write all messages to LOG_FILE
 * - 'none': No output (for performance testing)
 *
 * USAGE:
 * 1. Change OUTPUT_MODE constant below to desired mode
 * 2. Run: node swimmingpool.js
 * 3. Check results in console or LOG_FILE
 */

const SimLuxJS = require('../SimLuxJS/SimLuxJS.js').SimLuxJS; 
const SimEntity = require('../SimLuxJS/SimLuxJS.js').SimEntity;  
const seedrandom = require('seedrandom');
const fs = require('fs');

const OUTPUT_MODE = 'file'; // 'console', 'file', or 'none'
const LOG_FILE = 'simulation_js_output.log';

const RANDOM_SEED = 42;
const SIM_DURATION = 5 * 8 * 60;
const POOL_CAPACITY = 50;
const MAX_QUEUE_LENGTH = 30;
const NUMBER_SIM_EXPERIMENTS = 20;

// Logging function that respects OUTPUT_MODE
function logMessage(message) {
    switch(OUTPUT_MODE) {
        case 'console':
            console.log(message);
            break;
        case 'file':
            fs.appendFileSync(LOG_FILE, message + '\n');
            break;
        case 'none':
            // No output - for performance testing
            break;
    }
}

// Initialize log file (clear previous content if using file output)
function initializeLogging() {
    if (OUTPUT_MODE === 'file') {
        // Clear the log file at the start
        fs.writeFileSync(LOG_FILE, '');
        console.log(`Logging to file: ${LOG_FILE}`);
    } else if (OUTPUT_MODE === 'console') {
        console.log('Logging to console');
    } else {
        console.log('Logging disabled for performance testing');
    }
}

let random;
const uniform = (min, max) => random() * (max - min) + min;
const exponential = (lambda) => -Math.log(1 - random()) / lambda;

class Statistics {
    constructor() {
        this.waitingTimes = [];
        this.totalCustomers = 0;
        this.servedCustomers = 0;
    }

    recordWait(waitTime) {
        this.waitingTimes.push(waitTime);
    }

    report() {
        if (this.waitingTimes.length === 0) {
            logMessage("No waiting time data.");
            return;
        }
        const avg = this.waitingTimes.reduce((a, b) => a + b, 0) / this.waitingTimes.length;
        logMessage("\nWaiting Time Report:");
        logMessage(`- Average: ${avg.toFixed(2)} min`);
        logMessage(`- Max: ${Math.max(...this.waitingTimes).toFixed(2)} min`);
        logMessage(`- Min: ${Math.min(...this.waitingTimes).toFixed(2)} min`);
        logMessage(`- Total customers: ${this.totalCustomers}`);
        logMessage(`- Total served: ${this.servedCustomers}`);
        logMessage(`- Total waiting times recorded: ${this.waitingTimes.length}`);
        logMessage(`- Total waiting time: ${this.waitingTimes.reduce((a, b) => a + b, 0).toFixed(2)} min`);
        logMessage(`- Average waiting time per customer: ${(this.waitingTimes.reduce((a, b) => a + b, 0) / this.waitingTimes.length).toFixed(2)} min`);
        logMessage(`- Capacity: ${(this.waitingTimes.length / (SIM_DURATION / 60)).toFixed(2)} persons/hour`);
    }
}

class SwimmingPool {
    constructor(sim) {
        this.sim = sim;
        this.gateOpen = true;
        this.numInside = 0;
        this.numWaiting = 0;
        this.capacity = POOL_CAPACITY;
        this.stats = new Statistics();
    }

    canEnter() {
        return this.gateOpen && this.numInside < this.capacity;
    }

    addSwimmer() {
        this.numInside++;
    }

    removeSwimmer() {
        this.numInside--;
    }

    async openGateCycle() {
        while (this.sim.getTime() < SIM_DURATION) {
            this.gateOpen = true;
            logMessage(`\n[${this.sim.getTime().toString().padStart(5)}] GATE OPEN`);
            await this.sim.advance(1);
            this.gateOpen = false;
            logMessage(`[${this.sim.getTime().toString().padStart(5)}] GATE CLOSED`);
            await this.sim.advance(59);
        }
    }

    report() {
        this.stats.report();
    }
}

class Customer {
    static idCounter = 0;

    constructor(sim, pool) {
        Customer.idCounter++;
        this.name = `Swimmer ${Customer.idCounter}`;
        this.pool = pool;
        this.sim = sim;
        this.sim.addSimEntity(new SimEntity(simEntity => this.run()));
    }

    async run() {
        this.pool.numWaiting++;
        const waitStart = this.sim.getTime();
        logMessage(`[${this.sim.getTime().toString().padStart(5)}] ${this.name} arrives (waiting: ${this.pool.numWaiting}, inside: ${this.pool.numInside})`);

        while (!this.pool.canEnter()) {
            await this.sim.advance(1);
        }

        this.pool.numWaiting--;
        const waitEnd = this.sim.getTime();
        this.pool.stats.recordWait(waitEnd - waitStart);
        this.pool.addSwimmer();

        logMessage(`[${this.sim.getTime().toString().padStart(5)}] ${this.name} enters the pool (inside: ${this.pool.numInside})`);

        // Determine stay duration
        const w = random();
        let swimTime;
        if (w <= 0.6) {
            swimTime = uniform(115, 125); // ~2h +/- 5min
        } else {
            swimTime = uniform(75, 120);  // up to 45 min earlier
        }

        await this.sim.advance(swimTime);

        this.pool.removeSwimmer();
        this.pool.stats.servedCustomers++;
        logMessage(`[${this.sim.getTime().toString().padStart(5)}] ${this.name} leaves the pool (inside: ${this.pool.numInside})`);
    }
}

class ArrivalProcess {
    constructor(sim, pool) {
        this.pool = pool;
        this.sim = sim;
        this.sim.addSimEntity(new SimEntity(simEntity => this.run()));
    }

    async run() {
        while (this.sim.getTime() < SIM_DURATION) {
            await this.sim.advance(exponential(1)); // Mean interarrival: 1 min
            if (this.pool.numWaiting < MAX_QUEUE_LENGTH) {
                new Customer(this.sim, this.pool);
                this.pool.stats.totalCustomers++;
            }
        }
    }
}

async function runSingleExperiment(experimentNumber = 0) {
    // Set up random number generator with different seed for each experiment
    random = seedrandom(RANDOM_SEED + experimentNumber);
    
    Customer.idCounter = 0;
    let simLuxJS = new SimLuxJS();
    simLuxJS.enableLogging = false; // Disable logging for performance

    const pool = new SwimmingPool(simLuxJS);

    // Start arrival process
    new ArrivalProcess(simLuxJS, pool);

    // Start gate cycle process
    simLuxJS.addSimEntity(new SimEntity(simEntity => pool.openGateCycle()));

    logMessage(`\nRunning simulation experiment ${experimentNumber}...`);
    await simLuxJS.run(until=SIM_DURATION);
    logMessage(`\nSimulation ${experimentNumber} finished at ${simLuxJS.getTime()} minutes`);

    pool.report();

    return pool.stats;
}   

async function runAllExperiments() {
    initializeLogging();
    
    const totalTimes = [];
    for (let experiment = 1; experiment <= NUMBER_SIM_EXPERIMENTS; experiment++) {        
        const startTime = performance.now();
        await runSingleExperiment(experiment);
        const endTime = performance.now();

        const elapsedTime = endTime - startTime;
        totalTimes.push(elapsedTime);
    }
    
    const totalTime = totalTimes.reduce((a, b) => a + b, 0);
    const avgTime = totalTime / totalTimes.length;
    const minTime = Math.min(...totalTimes);
    const maxTime = Math.max(...totalTimes);
    
    console.log("\nSimLuxJS Performance Summary:");
    console.log(`- Average time: ${avgTime.toFixed(2)} ms`);
    console.log(`- Min time: ${minTime.toFixed(2)} ms`);
    console.log(`- Max time: ${maxTime.toFixed(2)} ms`);
    console.log(`- Total experiments: ${NUMBER_SIM_EXPERIMENTS}`);
    console.log(`- Total time: ${(totalTime / 1000).toFixed(2)} seconds`);
}

// Start the performance test
runAllExperiments().catch(console.error);

File: PerformanceTest/swimmingpool.py
"""
Swimming Pool Simulation (Python) - WITH LOGGING
====================================================
This script simulates a swimming pool environment using SimPy.
It models customer arrivals, swimming durations, and pool capacity constraints.
With configurable logging modes, it allows for testing and debugging.

LOGGING MODES:
- 'console': Real-time output to terminal
- 'file': Write all messages to LOG_FILE
- 'none': No output (for performance testing)

USAGE:
1. Change OUTPUT_MODE constant below to desired mode
2. Run: python swimmingpool.py
3. Check results in console or LOG_FILE
"""

import simpy
import random
import statistics
import time

OUTPUT_MODE = 'none'  # 'console', 'file', or 'none'
LOG_FILE = 'simulation_py_output.log'

RANDOM_SEED = 42
SIM_DURATION = 5 * 8 * 60 
POOL_CAPACITY = 100
MAX_QUEUE_LENGTH = 30
NUMBER_SIM_EXPERIMENTS = 20

# Logging function that respects OUTPUT_MODE
def log_message(message):
    if OUTPUT_MODE == 'console':
        print(message)
    elif OUTPUT_MODE == 'file':
        with open(LOG_FILE, 'a', encoding='utf-8') as f:
            f.write(message + '\n')
    # 'none' mode does nothing - for performance testing

# Initialize log file (clear previous content if using file output)
def initialize_logging():
    if OUTPUT_MODE == 'file':
        # Clear the log file at the start
        with open(LOG_FILE, 'w', encoding='utf-8') as f:
            f.write('')
        print(f"Logging to file: {LOG_FILE}")
    elif OUTPUT_MODE == 'console':
        print('Logging to console')
    else:
        print('Logging disabled for performance testing')

class Statistics:
    def __init__(self):
        self.waiting_times = []
        self.total_customers = 0
        self.served_customers = 0

    def record_wait(self, wait_time):
        self.waiting_times.append(wait_time)

    def report(self):
        if not self.waiting_times:
            log_message("No waiting time data.")
            return
        log_message("\nWaiting Time Report:")
        log_message(f"- Average: {statistics.mean(self.waiting_times):.2f} min")
        log_message(f"- Max: {max(self.waiting_times):.2f} min")
        log_message(f"- Min: {min(self.waiting_times):.2f} min")
        log_message(f"- Total waiting times recorded: {len(self.waiting_times)}")
        log_message(f"- Total waiting time: {sum(self.waiting_times):.2f} min")
        log_message("\nCustomer Report:")
        log_message(f"- Total customers: {self.total_customers}")
        log_message(f"- Total served customers: {self.served_customers}")
        log_message(f"- Capacity (customers/hour): {len(self.waiting_times) / (SIM_DURATION / 60):.2f} customers/hour")


class SwimmingPool:
    def __init__(self, env):
        self.env = env
        self.gate_open = True
        self.num_inside = 0
        self.num_waiting = 0
        self.capacity = POOL_CAPACITY
        self.stats = Statistics()

    def can_enter(self):
        return self.gate_open and self.num_inside < self.capacity

    def add_swimmer(self):
        self.num_inside += 1

    def remove_swimmer(self):
        self.num_inside -= 1

    def open_gate_cycle(self):
        while self.env.now < SIM_DURATION:
            self.gate_open = True
            log_message(f"\n[{self.env.now:>5}] GATE OPEN")
            yield self.env.timeout(1)
            self.gate_open = False
            log_message(f"[{self.env.now:>5}] GATE CLOSED")
            yield self.env.timeout(59)

    def report(self):
        self.stats.report()

class Customer:
    id_counter = 0

    def __init__(self, env, pool):
        Customer.id_counter += 1
        self.name = f"Swimmer {Customer.id_counter}"
        env.process(self.run(env, pool))

    def run(self, env, pool):
        pool.num_waiting += 1
        wait_start = env.now
        log_message(f"[{env.now:>5}] {self.name} arrives (waiting: {pool.num_waiting}, inside: {pool.num_inside})")

        while not pool.can_enter():
            yield env.timeout(1)

        pool.num_waiting -= 1
        wait_end = env.now
        pool.stats.record_wait(wait_end - wait_start)
        pool.add_swimmer()

        log_message(f"[{env.now:>5}] {self.name} enters the pool (inside: {pool.num_inside})")

        # Determine stay duration
        w = random.random()
        if w <= 0.6:
            swim_time = random.uniform(115, 125)  # ~2h +/- 5min
        else:
            swim_time = random.uniform(75, 120)   # up to 45 min earlier

        yield env.timeout(swim_time)

        pool.remove_swimmer()
        pool.stats.served_customers += 1
        log_message(f"[{env.now:>5}] {self.name} leaves the pool (inside: {pool.num_inside})")

def arrival_process(env, pool):
    while env.now < SIM_DURATION:
        yield env.timeout(random.expovariate(1))  # Mean interarrival: 1 min
        if pool.num_waiting < MAX_QUEUE_LENGTH:
            Customer(env, pool)
            pool.stats.total_customers += 1


def run_single_experiment(experiment_number=0):
    random.seed(RANDOM_SEED + experiment_number)
    Customer.id_counter = 0  
    env = simpy.Environment()
    pool = SwimmingPool(env)

    # Start arrival process
    env.process(arrival_process(env, pool))
    # Start gate cycle process
    env.process(pool.open_gate_cycle())
    
    log_message(f"Running simulation experiment {experiment_number}...")
    env.run(until=SIM_DURATION)
    log_message(f"\nSimulation {experiment_number} finished at {env.now} minutes")
    
    pool.report()
    if hasattr(env, 'reset'):
        env.reset()

def run_all_experiments():
    initialize_logging()

    total_times = []
    for experiment in range(1, NUMBER_SIM_EXPERIMENTS + 1):
        start_time = time.perf_counter()
        run_single_experiment(experiment)
        end_time = time.perf_counter()
        elapsed_time = (end_time - start_time) * 1000  # Convert to milliseconds
        total_times.append(elapsed_time)

    avg_time = sum(total_times) / len(total_times)
    min_time = min(total_times)
    max_time = max(total_times)
    
    print(f"\nSimPy Performance Summary:")
    print(f"- Average time: {avg_time:.2f} ms")
    print(f"- Min time: {min_time:.2f} ms")
    print(f"- Max time: {max_time:.2f} ms")
    print(f"- Total experiments: {NUMBER_SIM_EXPERIMENTS}")
    print(f"- Total time: {(sum(total_times) / 1000):.2f} seconds")

# Start the performance test
run_all_experiments()
